### ISSUE TITLE
[C#] Implement new Concept Exercise: string-formatting

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#962]
This issue describes how to implement the `string-formatting` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the basics of the Concept of String formatting in C#.

## Learning objectives

- Know how to use the `ToString()` method to convert any object to a `string`.
- Know how to use string interpolation on values of any type.
- Know how to use default format strings to convert to standard output formats.
- Know how to use custom format strings to convert to custom output formats.
- Know that `string.Format` underlies string interpolation.
- Know of the `StringBuilder` type and when to use it.
- Know that string interpolation can interpolate any expression.

## Out of scope

Nothing.

## Concepts

This Concepts Exercise's Concepts are:

- `string-formatting`: know how to use the `ToString()` method to convert any object to a `string`; know how to use string interpolation on values of any type; know how to use default format strings to convert to standard output formats; know how to use custom format strings to convert to custom output formats; know that `string.Format` underlies string interpolation; know of the `StringBuilder` type and when to use it; know that string interpolation can interpolate any expression.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `strings`: strings will be formatted.
- `inheritance`: knowing that each class derives from `object` and thus has built-in methods.

Any data types used in this exercise (e.g. `datetimes`) should also be added as prerequisites.

### Hints

- [String interpolation][string-interpolation]: tutorial on how to use string interpolation.
- [Formatting types][formatting-types]: the available format types to use.
- [Standard numeric format strings][standard-numeric-format-strings]: lists the standard numeric format strings.
- [Custom numeric format strings][custom-numeric-format-strings]: describes how to define custom numeric format strings.
- [Standard date and time format strings][standard-date-and-time-format-strings]: lists the standard date and time format strings.
- [Custom date and time format strings][custom-date-and-time-format-strings]: describes how to define custom date and time format strings.
- [String builder][string-builder]: shows how to use the string builder.

### After

- [String interpolation][string-interpolation]: tutorial on how to use string interpolation.
- [Formatting types][formatting-types]: the available format types to use.
- [Standard numeric format strings][standard-numeric-format-strings]: lists the standard numeric format strings.
- [Custom numeric format strings][custom-numeric-format-strings]: describes how to define custom numeric format strings.
- [Standard date and time format strings][standard-date-and-time-format-strings]: lists the standard date and time format strings.
- [Custom date and time format strings][custom-date-and-time-format-strings]: describes how to define custom date and time format strings.
- [String builder][string-builder]: shows how to use the string builder.
- [String interpolation - in-depth][string-interpolation-in-depth]: in-depth examination of string interpolation, including the generated IL code.
- [String interpolation - advanced][string-interpolation-advanced]: advanced tutorial on string interpolation, includes the `FormattableString` class.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[string-interpolation]: https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/string-interpolation
[string-interpolation-in-depth]: https://weblog.west-wind.com/posts/2016/Dec/27/Back-to-Basics-String-Interpolation-in-C#
[string-interpolation-advanced]: https://www.meziantou.net/interpolated-strings-advanced-usages.htm
[formatting-types]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/formatting-types
[standard-numeric-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings
[custom-numeric-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-numeric-format-strings
[standard-date-and-time-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings
[custom-date-and-time-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings
[string-builder]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/stringbuilder

**Comment by mikedamay** [comment]
Please have a look at the [instructions](https://github.com/mikedamay/v3/blob/csharp/string-formatting/languages/csharp/exercises/concept/string-formatting/.docs/instructions.md).

This is based on _Elixir's_ `strings` exercise.

Tasks 1 and 3 push towards string interpolation.

Task 2 pushes towards `string.Format()` and verbatim strings.

**Comment by ErikSchierboom** [comment]
Yes, I like it! 👍

**Comment by mikedamay** [comment]
closed with PR #1871
